### ISSUE TITLE
Incompatible type error when creating a list from a static method or class method returning an iterator of namedtuples

### ISSUE THREAD

**Issue by mikez302** [python/mypy#4546]
I am using Python 3.6.4 and mypy 0.560. I have a file with the following code:

```python
from collections import namedtuple
from typing import Iterator, List


class Company:
    def get_person_list(self) -> List['Person']:
        return list(self.get_people())

    @staticmethod
    def get_people() -> Iterator['Person']:
        yield Person('Alice', 20)
        yield Person('Bob', 50)


class Person(namedtuple('Person', ['name', 'age'])):
    pass
```

If I run it through mypy, I get an error: `iterator_test2.py:7: error: Argument 1 to "set" has incompatible type "Iterator[Person]"; expected "Iterable[Person]"`

I don't know why this is happening. It looks like something is wrong with the type of the iterator I'm passing to the list constructor.

But here is where it gets weird:
* If I remove either of the return type hints from the methods in `Company`, I get no errors.
* If I move the `Person` class above the `Company` class, I get no errors.
* If I make `get_people` a regular method instead of a static method (and add a `self` parameter), I get no errors.
* If I move `get_people` outside of `Company` (making it a regular function instead of a static method, and removing the `@staticmethod` decorator), I get no errors.
* If I make `get_people` yield ints instead of Persons, and change the annotations to `List[int]` and `Iterator[int]`, I get no errors.
* If I make `get_people` yield regular tuples instead of namedtuples (and change the return type hints in the methods of `Company` accordingly), I get no errors.
* If I make `Person` a class that doesn't inherit from `namedtuple` (and sets its attributes in `__init__` like classes usually do), I get no errors (regardless of whether I have any type hints in `__init__`).

On the other hand:
* If I make `get_people` a class method (and add a `cls` parameter), I get the same error.
* If I make `get_people` inherit from `typing.NamedTuple('Person', [('name', str), ('age', int)])`, I get the same error.
* If I create the `Person` class with the statement `Person = namedtuple('Person', ['name', 'age'])`, I get this output:
    ```
    iterator_test2.py:7: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues version: 0.560
    iterator_test2.py:7: : note: please use --show-traceback to print a traceback when reporting a bug
    ```
    If I run mypy with `--show-traceback`, I get this output:
    ```
    iterator_test2.py:7: error: INTERNAL ERROR -- please report a bug at https://github.com/python/mypy/issues version: 0.560
    Traceback (most recent call last):
      File "/Library/Frameworks/Python.framework/Versions/3.6/bin/mypy", line 11, in <module>
        sys.exit(console_entry())
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/__main__.py", line 7, in console_entry
        main(None)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/main.py", line 66, in main
        res = type_check_only(sources, bin_dir, options)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/main.py", line 119, in type_check_only
        options=options)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/build.py", line 218, in build
        graph = dispatch(sources, manager)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/build.py", line 1997, in dispatch
        process_graph(graph, manager)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/build.py", line 2299, in process_graph
        process_stale_scc(graph, scc, manager)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/build.py", line 2475, in process_stale_scc
        graph[id].type_check_first_pass()
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/build.py", line 1876, in type_check_first_pass
        self.type_checker().check_first_pass()
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 196, in check_first_pass
        self.accept(d)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 286, in accept
        stmt.accept(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/nodes.py", line 696, in accept
        return visitor.visit_class_def(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 1210, in visit_class_def
        self.accept(defn.defs)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 286, in accept
        stmt.accept(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/nodes.py", line 756, in accept
        return visitor.visit_block(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 1330, in visit_block
        self.accept(s)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 286, in accept
        stmt.accept(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/nodes.py", line 497, in accept
        return visitor.visit_func_def(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 534, in visit_func_def
        self.check_func_item(defn, name=defn.name())
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 599, in check_func_item
        self.check_func_def(defn, typ, name)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 762, in check_func_def
        self.accept(item.body)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 286, in accept
        stmt.accept(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/nodes.py", line 756, in accept
        return visitor.visit_block(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 1330, in visit_block
        self.accept(s)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 286, in accept
        stmt.accept(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/nodes.py", line 869, in accept
        return visitor.visit_return_stmt(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 2101, in visit_return_stmt
        self.check_return_stmt(s)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checker.py", line 2132, in check_return_stmt
        allow_none_return=allow_none_func_call)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checkexpr.py", line 2365, in accept
        typ = node.accept(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/nodes.py", line 1245, in accept
        return visitor.visit_call_expr(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checkexpr.py", line 268, in visit_call_expr
        ret_type = self.check_call_expr_with_callee_type(callee_type, e, fullname, object_type)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checkexpr.py", line 505, in check_call_expr_with_callee_type
        object_type=object_type)[0]
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checkexpr.py", line 611, in check_call
        object_type=object_type)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checkexpr.py", line 578, in check_call
        messages=arg_messages)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checkexpr.py", line 1048, in check_argument_types
        actual + 1, i + 1, callee, context, messages)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/checkexpr.py", line 1080, in check_arg
        elif not is_subtype(caller_type, callee_type):
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/subtypes.py", line 83, in is_subtype
        ignore_declared_variance=ignore_declared_variance))
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/types.py", line 475, in accept
        return visitor.visit_instance(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/subtypes.py", line 169, in visit_instance
        zip(t.args, right.args, right.type.defn.type_vars))
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/subtypes.py", line 168, in <genexpr>
        for lefta, righta, tvar in
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/subtypes.py", line 38, in check_type_parameter
        return is_subtype(lefta, righta, check_type_parameter)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/subtypes.py", line 83, in is_subtype
        ignore_declared_variance=ignore_declared_variance))
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/types.py", line 1415, in accept
        return visitor.visit_forwardref_type(self)
      File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/mypy/types.py", line 1500, in visit_forwardref_type
        raise RuntimeError('Internal error: unresolved forward reference')
    RuntimeError: Internal error: unresolved forward reference
    iterator_test2.py:7: : note: use --pdb to drop into pdb
    ```
* If I create the `Person` class using the statement `Person = typing.NamedTuple('Person', [('name', str), ('age', int)])`, I get a similar error and a similar traceback.

I think I found a bug in mypy. My first example should be valid. I can post all of my examples of code that worked and all of the tracebacks here if anyone wants.

**Comment by gvanrossum** [comment]
Can you try your various cases again with master from the repo? I see the same error you quote with the code you post with v0.560, but not with master.

**Comment by mikez302** [comment]
@gvanrossum, I installed the latest version of mypy by running `python3 -m pip install -U git+git://github.com/python/mypy.git` from the command line, as described in the readme. I ran `mypy --version` and it told me `mypy 0.570-dev-26b51e554d7426a0aed6ad55bae7b10a4619c093`.

I tried my example, and all of the cases in both of my lists, and none of them produced any errors.

I guess this must have been fixed in master. Feel free to close this issue.
